### ISSUE TITLE
Enable Kafka Streams binder for Graal native compilation

### ISSUE THREAD

**Issue by sobychacko** [spring-cloud/spring-cloud-stream#2323]
@sobychacko commented on [Tue Apr 27 2021](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/1064)

---

@sabbyanandan commented on [Tue May 04 2021](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/1064#issuecomment-832023131)

Verification on the KTable is still pending.

**Comment by onobc** [comment]
Most likely this is blocked by the JVM reachability for the state store JNI as described in https://github.com/spring-projects-experimental/spring-native/issues/680

**Comment by sdeleuze** [comment]
My tests on https://github.com/spring-projects/spring-aot-smoke-tests/tree/main/cloud-stream-kafka seems to indicate it is now working, should this issue be updated accordingly?

**Comment by onobc** [comment]
@sdeleuze that AOT smoke test only covers the Kafka binder, not the KafkaStreams binder that this is referring to (naming... they sound very much like "the thing"). 

Once this issue is addressed (or it is no longer an issue) then another AOT smoke test should be added to cover it as well.

**Comment by pdebuitlear** [comment]
Support for Kafka Streams doesn't work out of the box, some additional hints are needed. See [this project](https://github.com/pdebuitlear/kstreams-demo/tree/main/src/main/java/com/example/demo/config/hints)  for a working example. There are also hints in that project for the Confluent schema registry and the Confluent serializers.

**Comment by sobychacko** [comment]
@pdebuitlear We haven't addressed native changes for the Kafka Streams binder yet, but we will do it as part of one of the upcoming `4.x` releases. Thanks for sharing the project, as that will be a good reference.

**Comment by pdebuitlear** [comment]
> Looking to that project, @pdebuitlear , I would say that I don't see anything Spring Cloud Stream related or even Spring at all. It really would be better to have Native images support directly from those libs, so any target implementation would benefit from a common place, not just this Spring Cloud Stream Kafka Streams Binder. So, I would suggest to consider raising improvement tickets in respective Kafka Streams and Confluent Platform projects respectively.

Not even these?
 https://github.com/pdebuitlear/kstreams-demo/blob/main/src/main/java/com/example/demo/config/hints/KafkaStreamsHints.java

**Comment by garyrussell** [comment]
Spring for Apache Kafka added a few hints for basic Kafka, and Kafka Streams, support, but this was for expediency, to prove out the core project can support native images; it is impractical for Spring projects to maintain ongoing, extensive, hints for all dependencies.

There are some sample smoke tests for the basic hints. https://github.com/spring-projects/spring-aot-smoke-tests/tree/main/integration

**Comment by sabbyanandan** [comment]
@sobychacko @artembilan @garyrussell @olegz: Many thanks for continuing to evolve the binders and making Kafka integration easier for Java developers. The developer community appreciates your time and commitment. 🙏🏽

**Comment by pdebuitlear** [comment]
> Spring for Apache Kafka added a few hints for basic Kafka, and Kafka Streams, support, but this was for expediency, to prove out the core project can support native images; it is impractical for Spring projects to maintain ongoing, extensive, hints for all dependencies.
> 
> There are some sample smoke tests for the basic hints. https://github.com/spring-projects/spring-aot-smoke-tests/tree/main/integration

That's a fair point, thanks. Does that leave the Spring stack gapped in comparison with other frameworks?

**Comment by artembilan** [comment]
I don’t find it as a correct comparison: such a native hint really has to be in that library, not any framework. This way all the frameworks and any library consumers would benefit. Not sure why you don’t see it as so obvious reasonable solution. Native images is really not Spring feature, so exposing hints for third parties is doubtful work for Spring. We cannot judge other framework solutions, but it is not so convenient to keep hints for common libs just in Spring.

**Comment by pdebuitlear** [comment]
> I don’t find it as a correct comparison: such a native hint really has to be in that library, not any framework. This way all the frameworks and any library consumers would benefit. Not sure why you don’t see it as so obvious reasonable solution. Native images is really not Spring feature, so exposing hints for third parties is doubtful work for Spring. We cannot judge other framework solutions, but it is not so convenient to keep hints for common libs just in Spring.

Ok forget the Confluent stuff for the moment, I can create a project in Quarkus for Kafka Streams and it will add hints for the Apache Kafka classes and dependent RocksDB .so libs. That does not work with Spring Native/Aot code currently.

**Comment by artembilan** [comment]
What if I don’t want Quarkus and just want to build native app just with plain Kafka Streams? Why would I always go for Quarkus just because it exposes hints for RocksDB? 

I find your push to us as counterproductive: the native hint is really has to be exposed in that library which is responsible for those types. Is it Quarkus who maintains RocksDB? Why similar native hints have to be in Quarkus and Spring? Why just Apache Kafka cannot provide for us all a single solution for types it supports? 
I just really think here about broader Apache Kafka and GraalVM community, not just those who make their applications based on Spring.

**Comment by pdebuitlear** [comment]
[_"Kafka Streams uses RocksDB as the default storage engine for persistent stores"_](https://kafka.apache.org/24/documentation/streams/developer-guide/config-streams)  It's a dependent jar of Apache Kafka. It's just a bit strange that some functionality in Apache Kafka would have 'hints' and some wouldn't. At least document it, if your going to make life difficult for your user base.

**Comment by sdeleuze** [comment]
- Initial intent should be to contibute hints and related tests to the various libraries.
- If they refuse for some reasons, remaining non Spring related hints should be contributed to https://github.com/oracle/graalvm-reachability-metadata, where Spring is the #1 contributor by far.
- Spring related hints should live in respective Spring projects.

Happy to help on contributions on https://github.com/oracle/graalvm-reachability-metadata if needed.

**Comment by sdeleuze** [comment]
Could you try to contribute the third party bits to https://github.com/oracle/graalvm-reachability-metadata like we do for the rest of the Spring portfolio?

**Comment by pdebuitlear** [comment]
> Could you try to contribute the third party bits to https://github.com/oracle/graalvm-reachability-metadata like we do for the rest of the Spring portfolio?

@sdeleuze was that directed to myself?

**Comment by sobychacko** [comment]
@pdebuitlear We can look into this as a start. Could you please describe your use case once again? Are you trying to use the Kafka Streams binder in Spring Cloud Stream with graalvm? Is that where you run into the issues?

**Comment by sdeleuze** [comment]
> @sdeleuze was that directed to myself?

Not specifically

**Comment by StevenPG** [comment]
Does the presence of spring-cloud-streams kafka-streams within the smoke tests repository imply that it should be working in some capacity?

https://github.com/spring-projects/spring-aot-smoke-tests/tree/main/cloud/cloud-stream-kafka

We also are intending to run spring-cloud-streams applications using the kafka-streams binder using native compilation.

**Comment by sobychacko** [comment]
@0xabadea As soon as the Spring One conference finishes next week, this issue will be a priority for us, and we will work on it. Please stay tuned.

**Comment by rob-64** [comment]
@sobychacko Sorry to bother you, but is there any ETA or progress on this?

**Comment by sobychacko** [comment]
We are targeting this as part of the `RC1` release scheduled for the end of this month.

**Comment by sobychacko** [comment]
The basic support for enabling Kafka Streams binder for native applications is in place. cc @rob-64 @0xabadea @andrashatvani

**Comment by rob-64** [comment]
@sobychacko is there a snapshot repo where this is published?

**Comment by onobc** [comment]
Hi @rob-64 




You can find snapshots on repo.spring.io/snapshot (ex: https://repo.spring.io/ui/native/snapshot/org/springframework/cloud/spring-cloud-stream-binder-kafka-streams/4.1.0-SNAPSHOT/)

Easiest way to setup for this is to go to start.spring.io and choose a snapshot Boot version. The resulting build (Maven or Gradle) will be setup accordingly.

**Comment by marcoberri** [comment]
Hi,
I try with:
- spring-boot-starter-parent : 3.2.0-RC1 and 3.0.1
- https://github.com/pdebuitlear/kstreams-demo/blob/main/src/main/java/com/example/demo/config/hints/KafkaStreamsHints.java
- spring-cloud-stream-binder-kafka:4.1.0-SNAPSHOT
 
when run a exe: 

`10:40:59.658 ERROR --- [           main] o.s.boot.SpringApplication               : Application run failed

java.lang.NoSuchMethodError: java.lang.invoke.MethodHandle.get()
        at org.springframework.boot.autoconfigure.websocket.servlet.WebSocketServletAutoConfiguration__BeanDefinitions$JettyWebSocketConfiguration.getJettyWebSocketConfigurationBeanDefinition(WebSocketServletAutoConfiguration__BeanDefinitions.java:30)
`
exists any solution?

tnx

**Comment by sobychacko** [comment]
@marcoberri That looks unrelated to the Kafka Streams binder. [Here](https://github.com/spring-projects/spring-aot-smoke-tests/tree/main/cloud/cloud-stream-kafka-streams) is a sample app and the smoke test for this in the official spring-aot-smoke-tests repository. Please compare it with your setup. If you still have issues, please create a new issue and provide a simple sample app (without any external dependencies) so we can triage further. Thank you.

**Comment by marcoberri** [comment]
thanks @sobychacko, it's partly solved. Now I still have this problem in execution phase:

`java.lang.UnsatisfiedLinkError: Can't load library: /tmp/librocksdbjni5951241590552195586.so
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.NativeLibrarySupport.loadLibraryAbsolute(NativeLibrarySupport.java:100)
`

**Comment by sobychacko** [comment]
@marcoberri The hints for the rocksdb jni libraries are there in the [binder](https://github.com/spring-cloud/spring-cloud-stream/blob/main/binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/aot/KafkaStreamsBinderRuntimeHints.java#L69).  Are you using the latest snapshots? Please open a new issue and provide a small sample app where we can reproduce the error.

**Comment by marcoberri** [comment]
@sobychacko 
I think the problem is accesso to java.library.path. The library exists in docker image

![image](https://github.com/spring-cloud/spring-cloud-stream/assets/1119268/736d5cf9-64aa-4f3d-ae43-2c95dfd102e3)

if I try to print the variable  java.library.path in running application i see:  Library Path :/usr/lib64:/lib64:/lib:/usr/lib:/tmp
the /tmp folder configuration exists

but when run: 
java.lang.UnsatisfiedLinkError: Can't load library: /tmp/librocksdbjni421600095650118164.so
